File: __prologue_server/mapGen.js
/* */

import { aux } from "./lib/aux.js";
import { K } from "./_konfig.js";

function generate(typeOfMap, cols, rows, rnd) {
  switch (typeOfMap) { // outdoors,indoors
    case "basicRoom":
      return basicRoom.create(cols, rows, rnd);
    case "shelter":
      return shelter.create();
  }
}

export {
  generate,
};

//////////

class Tile {
  constructor(terrain) {
    this.createTile(terrain);
  }
  createTile(terrain) {
    let p = [];
    switch (terrain) {
      case "floor":
        p = [terrain, true, false, false, false];
        break;
      case "wall":
        p = [terrain, false, true, false, false];
        break;
      case "unknown":
        p = [terrain, false, true, false, false];
        break;
      default:
        p = [];
    }
    this.terrain = terrain;
    this.walkable = p[1];
    this.blockLOS = p[2];
    this.explored = p[3];
    this.visible = p[4];
  }
}

///////////////////////////////
/////////// SHELTER ///////////
///////////////////////////////

const shelter = {
  create: function name(cols, rows) {
    return basicRoom.create(cols, rows);
  }
};

///////////////////////////////
///////// BASIC ROOM //////////
///////////////////////////////

const basicRoom = {
  map: [],
  create: function name(cols, rows, rnd) {
    const columns = Math.floor((cols * rows) / 30);
    this.map = aux.initializeMultiArray(cols, rows, {});
    this.clean(cols, rows);
    this.putColumns(cols, rows, columns, rnd);
    return this.map;
  },
  clean: function (cols, rows) {
    for (let x = 0; x < cols; x++) {
      for (let y = 0; y < rows; y++) {
        if (x !== 0 && x !== cols - 1 && y !== 0 && y !== rows - 1) {
          this.map[x][y] = new Tile("floor");
        } else {
          this.map[x][y] = new Tile("wall");
        }
      }
    }
  },
  putColumns: function (cols, rows, columns, rnd) {
    let tries = 0;
    while (columns > 0 && tries < K.tries) {
      const x = rnd.int(1, cols - 2);
      const y = rnd.int(1, rows - 2);
      if (this.map[x][y].terrain === "floor") {
        this.map[x][y] = new Tile("wall");
        columns--;
      }
      tries++;
    }
  },
};

File: __prologue_server/mapGen.go
/* */

package main

import (
	"math/rand"
	"prologue/lib"
)

type Level [][]Tile

type Tile struct {
	Terrain  string `json:"terrain"`
	Walkable bool   `json:"walkable"`
	BlockLOS bool   `json:"blockLOS"`
	Explored bool   `json:"explored"`
	Visible  bool   `json:"visible"`
}

func (t Tile) create(terrain string) Tile {
	switch terrain {
	case "floor":
		return Tile{terrain, true, false, false, false}
	case "wall":
		return Tile{terrain, false, true, false, false}
	case "unknown":
		return Tile{terrain, false, true, false, false}
	}
	return Tile{}
}

func NewLevel(lvlType string, x *rand.Rand, cols, rows int) Level {
	switch lvlType {
	case "testRoom":
		return testRoom(x, cols, rows)
	}
	return nil
}

///////////////////////////////
////////// TEST ROOM //////////
///////////////////////////////

func testRoom(x *rand.Rand, cols, rows int) Level {
	lvl := createTestRoom(cols, rows)
	lvl = fillTestRoomWith(lvl, "floor", cols, rows)
	lvl = cleanTestRoom(lvl, "wall", cols, rows)
	lvl = putColumnsTestRoom(lvl, 50, x, cols, rows)
	return lvl
}

func putColumnsTestRoom(lvl Level, many int, rnd *rand.Rand, cols, rows int) Level {
	for c := 0; c < many; c++ {
		x := lib.RandomInt(1, cols-1, rnd)
		y := lib.RandomInt(1, rows-1, rnd)
		if lvl[x][y].Walkable {
			lvl[x][y] = Tile{}.create("wall")
		}
	}
	return lvl
}

func cleanTestRoom(lvl Level, fill string, cols, rows int) Level {
	for x := 0; x < cols; x++ {
		for y := 0; y < rows; y++ {
			if y == 0 || x == 0 || x == cols-1 || y == rows-1 {
				lvl[x][y] = Tile{}.create(fill)
			}
		}
	}
	return lvl
}

func createTestRoom(cols, rows int) Level {
	lvl := make([][]Tile, cols)
	for i := range lvl {
		lvl[i] = make([]Tile, rows)
	}
	for x := 0; x < cols; x++ {
		for y := 0; y < rows; y++ {
			lvl[x][y] = Tile{}
		}
	}
	return lvl
}

func fillTestRoomWith(lvl Level, fill string, cols, rows int) Level {
	for x := 0; x < cols; x++ {
		for y := 0; y < rows; y++ {
			lvl[x][y] = Tile{}.create(fill)
		}
	}
	return lvl
}
